legate/core/runtime.py

```diff
@@ -131,12 +131,18 @@ def update(self, **kwargs: Any) -> None:
     def remove(self, key: str) -> None:
         del self._entries[key]
 
-    def __repr__(self) -> str:
+    def machine_string(self) -> str:
         pairs = [f"{key},{value}" for key, value in self._entries.items()]
         if self._provenance is not None:
             pairs.append(f"Provenance,{self._provenance}")
         return "|".join(pairs)
 
+    def user_string(self) -> str:
+        pairs = [f"{key}={value}" for key, value in self._entries.items()]
+        return (
+            "<unknown>" if self._provenance is None else self._provenance
```

Comment:
no right or wrong way, but I find the following more readable:

```
provenance_str = "<unknown>" if self._provenance is None else self._provenance
if pairs:
  return provenance_str + f"[{','.join(pairs)}]"
return provenance_str
```

Reply:
TBH I would probably just live with `[]` in the empty case for simplicity:
```
return (self._provenance or "<unknown>") + + f"[{','.join(pairs)}]"
```

Reply:
Or if you really want it to look nice in that case:
```python
entries = ",".join(f"{k}={v}" for k, v in self._entries.items())
provenance = self._provenance or "<unknown>"
return provenance + f" [{entries}]" if entries else ""
```